janus_server/src/hpke.rs

```diff
@@ -0,0 +1,379 @@
+                        self.task_id,
+                        self.label,
+                        self.sender_role,
+                        self.recipient_role,
+                    ),
+                    plaintext,
+                    associated_data,
+                )
+            }
+            (HpkeKemId::X25519HkdfSha256, HpkeKdfId::HkdfSha512, HpkeAeadId::ChaCha20Poly1305) => {
+                seal::<ChaCha20Poly1305, HkdfSha512, X25519HkdfSha256>(
+                    &self.recipient_config,
+                    HpkeApplicationInfo::new(
+                        self.task_id,
+                        self.label,
+                        self.sender_role,
+                        self.recipient_role,
+                    ),
+                    plaintext,
+                    associated_data,
+                )
+            }
+            (_, _, _) => Err(Error::InvalidConfiguration(
+                "unsupported set of HPKE algorithms",
+            )),
+        }
+    }
+}
+
+// This function exists separately from struct HpkeSender to abstract away its
+// generic parameters
+fn seal<Encrypt: Aead, Derive: Kdf, Encapsulate: Kem>(
+    recipient_config: &HpkeConfig,
+    application_info: HpkeApplicationInfo,
+    plaintext: &[u8],
+    associated_data: &[u8],
+) -> Result<HpkeCiphertext, Error> {
+    let mut rng = thread_rng();
+
+    // Deserialize recipient pub into the appropriate PublicKey type for the
+    // KEM
+    let recipient_public_key = Encapsulate::PublicKey::from_bytes(&recipient_config.public_key.0)?;
+
+    let (encapsulated_context, mut context) = setup_sender::<Encrypt, Derive, Encapsulate, _>(
+        &OpModeS::Base,
+        &recipient_public_key,
+        &application_info.0,
+        &mut rng,
+    )?;
+
+    Ok(HpkeCiphertext {
+        config_id: recipient_config.id,
+        encapsulated_context: encapsulated_context.to_bytes().to_vec(),
+        payload: context.seal(plaintext, associated_data)?,
+    })
+}
+
+/// An HPKE recipient that decrypts messages encrypted to the public key in
+/// `recipient_config`, using the AEAD, key derivation and key encapsulation
+/// mechanisms specified in `recipient_config`, and using `label`, `sender_role`
+/// and `recipient_role` to derive application info.
+//
+// This type only exists separately from Recipient so that we can have a type
+// that doesn't "leak" the generic type parameters into the caller.
+struct HpkeRecipient {
+    task_id: TaskId,
+    config: HpkeConfig,
+    label: Label,
+    sender_role: Role,
+    recipient_role: Role,
+    recipient_private_key: HpkePrivateKey,
+}
+
+impl HpkeRecipient {
+    /// Generate a new X25519HkdfSha256 keypair and construct an HPKE recipient
+    /// using the private key, with KEM = X25519HkdfSha256, KDF = HkdfSha512 and
+    /// AEAD = ChaCha20Poly1305, and the specified label, and roles.
+    fn generate(label: Label, sender_role: Role, recipient_role: Role) -> Self {
```

Comment:
I presume this method is just for testing purposes, right? (based on the random task_id, zero HpkeConfig.id, and ephemeral keypairs)

Reply:
I think so; if so, I suggest moving this to the `tests` module, or otherwise arranging for this to be `#[cfg(test)]`.

Reply:
We will need this method outside of test targets in an upcoming PR, which will create an `aggregator` binary that uses this to implement an `hpke_config` endpoint. Past that we'll have to figure out how to load HPKE keys and configurations from config, but we eventually need to generate keys which might use a modified version of this method.

All that being said, it makes sense to restrict this method to `cfg(test)` in this PR.